### PR TITLE
Full Review

### PR DESCRIPTION
**Pull request by mr-joshcrane**
A full review of the entire repository. When this PR is complete, be sure to manually merge its head branch into the main branch for this repository.

### CODE REVIEW

File: remember.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		}
+		memorable = append(memorable, string(word))
+	}
+	data := strings.Join(memorable, " ")
+	data = data + "\n"
+	_, err := m.output.Write([]byte(data))
+	return err
+}
+
+func Reminder() string {
```
Comment (ivanfetch) [inline remember.go]: This is a nit-pick: I feel the function name `Reminder` doesn't fully match its dual purposes of memorizing and recalling from the data file. This is a good function to be calling from `main.go` (as you are doing) - I might call it `RunReminder` or similar, to help denote it is an entrypoint into your `remember` package.

File: remember.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		return nil
+	}
+}
+
+func NewMemory(opts ...Option) (Memory, error) {
+	_, err := os.Stat("store.txt")
+	if err != nil {
+		os.Create("store.txt")
+	}
+	defaultFile, err := os.OpenFile("store.txt", os.O_APPEND|os.O_RDWR, 0644)
```
Comment (ivanfetch) [inline remember.go]: This error isn't checked, in an `if err != nil` statement, which can kick the error condition down the road. For example, if this `os.OpenFile()` call fails due to an uncaught `open store.txt: permission denied` error, the later call to `ioutil.ReadAll()` returns the less useful error `error opening data store: invalid argument`.

Comment (ivanfetch) [inline remember.go]: This is a nit-pick: the variable name `defaultFile` makes me think there is (or will be) a non-default instance elsewhere. I'm not sure `default` should be part of the name, but there may be a purpose I don't yet understand.

File: remember.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			return Memory{}, err
+		}
+	}
+	return m, nil
+}
+
+func (m *Memory) Recall() string {
+	data, err := ioutil.ReadAll(m.output)
+	if err != nil {
+		panic("error opening data store")
```
Comment (ivanfetch) [inline remember.go]: This panic message could also include the reason for the open failure (like permission denied) from the above error, using `err.Error()` to get its message as a string.
Also, I think calling `panic()` here is reasonable if this is both a command-line specific function, and an unrecoverable condition. However, if a user of your package were creating their own command-line interface, and calling your `Recall()` function, they might want to handle an error reading the content of the data file.

File: remember.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func NewMemory(opts ...Option) (Memory, error) {
+	_, err := os.Stat("store.txt")
+	if err != nil {
+		os.Create("store.txt")
+	}
+	defaultFile, err := os.OpenFile("store.txt", os.O_APPEND|os.O_RDWR, 0644)
+	m := Memory{
+		[]io.Reader{os.Stdin},
```
Comment (ivanfetch) [inline remember.go]: I think specifying the struct fields by name, helps make assignments more understandable. For example:
```
	m := Memory{
		input:  []io.Reader{os.Stdin},
		output: defaultFile,
	}
```

File: remember.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	return func(m *Memory) error {
+		m.output = bytes.NewBufferString(output)
+		return nil
+	}
+}
+
+func NewMemory(opts ...Option) (Memory, error) {
+	_, err := os.Stat("store.txt")
+	if err != nil {
+		os.Create("store.txt")
```
Comment (ivanfetch) [inline remember.go]: `os.Create()` can also return an error, which is not being collected and checked here. Currently if `Create()` encounters an error, the first error a user will see is from `ioutil.ReadAll()` potentially failing to read data from a file that may have failed creation. This is similar to my other comment about the potential `os.OpenFile` error being kicked down the road.

File: remember.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+		err := opt(&m)
+		if err != nil {
+			return Memory{}, err
+		}
+	}
+	return m, nil
+}
+
+func (m *Memory) Recall() string {
+	data, err := ioutil.ReadAll(m.output)
```
Comment (ivanfetch) [inline remember.go]: After Go version 1.15, `ReadAll` is now in the `io` package. [io.ReadAll documentation](https://pkg.go.dev/io/ioutil#ReadAll) states "As of Go 1.16, this function simply calls io.ReadAll."

Comment (ivanfetch) [inline remember.go]: Using ReadAll will allocate memory to hold the entire file being read. This is alright if you know input will not be large. IF you are concerned about a user exhausting memory by providing a large input file, you can use `io.Copy()` to move the data in chunks between input and output `io.Reader`s.

File: cmd/main.go [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package main
+
+import (
+	"fmt"
+	"remember"
+)
+
+func main() {
+	fmt.Println(remember.Reminder())
```
Comment (ivanfetch) [inline cmd/main.go]: This is a nice short `main.go`. :)

File: remember_test.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
+package remember_test
+
+import (
+	"remember"
+	"testing"
+)
+
+func TestMemorise(t *testing.T) {
+	memory, err := remember.NewMemory(
```
Comment (ivanfetch) [inline remember_test.go]: This test can also be run in parallel with `t.Parallel()`, since it has its own `Memory` and input/output.

**Review by ivanfetch — CHANGES_REQUESTED** [review]
Hi Josh,

I like how you're starting off this project to include tests with good test-cases, like successive reminder entries. I also like the design-decisions that support test input/output, like using `io` interfaces, instead of reading/writing directly to a terminal from your package functions. Good stuff!

I made a few suggestions around error checking, and a few nit-picks which may not apply as I'm not aware of the future design and plans you have for this project.

Please feel free to ask me here or in Slack if you have any questions about any of the comments I've made. Also feel free to give me feedback about how this style of review works for you.


* I see you (or someone) used my pr-me tool to create a full review pull request of this repository. I want to call out an issue with my tool, to help make sure you do not lose any successive changes you may make to the `prme-*` branches:
  * When this pull request gets merged, any changes you may have committed to the `prme-full-content` branch are merged into the `prme-full-review` branch, instead of your `main` branch as is typical. This is due to git funkiness used to manufacture a "full pull request" to include 100% of the repo content. IF you happen to commit code to the `prme-full-content` branch, be sure to manually merge your changes into the `main` branch when this pull request is complete.

File: remember.go (hunk 2 above)
Comment (mr-joshcrane) [reply remember.go]: Now that I look at this with fresh eyes, I think my `NewMemory` function is overloaded, I do two things in it :|
I open the data store and I create the memory hrmm

File: remember.go (hunk 6 above)
Comment (mr-joshcrane) [reply remember.go]: That is a useful thing to know, I will keep it in mind in the future. However for this use case I am happy too offload the responsibility to the user on this one for now. 

Having said that, it does raise some interesting questions about input validation, ie what if a user mistakenly passed a binary fill and filled their datastore with junk

**Pull request closed without merging**